Rearchitect and Reinstate GPU

onurulgen opened the pull request:
-	Reinstated GPU into the project by fixing or implementing missing functions.
-	Rearchitected the project to combine GPU and CPU implementations by using virtual functions.
-	Increased the performance up to 11 times by using GPU.
-	Implemented unit and regression tests for GPU and CPU by using Catch2 framework.
-	Developed GitHub Actions for tests, coverage, static code analysis, and producing CPU and GPU executables for Windows, Linux and macOS.

onurulgen committed:
Add NiftiImage::realloc() to reallocate the image data

onurulgen committed:
Add NiftiImage::setDim() to set a dimension of the image

onurulgen committed:
Add ability to NiftiImage for copying only image info

onurulgen committed:
Update tests to leverage new abilities of NiftiImage

onurulgen committed:
Add an enum for NiftiImage dimensions

onurulgen committed:
Small fixes

onurulgen committed:
Add ability to NiftiImageData for extracting volume data

onurulgen committed:
Handle optimise* variables in Compute::NormaliseGradient()

onurulgen committed:
Refactorisations

onurulgen committed:
Add NiftiImage::copyData() to copy the pixel data from another image

onurulgen committed:
Extend NiftiImage constructor to optionally allocate image data

onurulgen committed:
Add NiftiImage::totalBytes() to return the total size of the image data in bytes

onurulgen committed:
Add move constructor and assignment operator to NiftiImageData

onurulgen committed:
Remove nodeNumber parameter from GetMaximalLength() and NormaliseGradient() of *Compute classes

onurulgen committed:
Fix possible bugs causing accessing freed memory

onurulgen committed:
Add tests for *Compute::GetMaximalLength() and *Compute::NormaliseGradient()

onurulgen committed:
Fix a bug in tests causing wrong voxel count calculation

onurulgen committed:
Fix HAVE_ZLIB definition

onurulgen committed:
Refactor CMakeLists

onurulgen committed:
Refactorisations

onurulgen committed:
Add NormaliseGradient() for CUDA and refactor GetMaximalLength()

onurulgen committed:
Extend tests for *Compute::GetMaximalLength() and *Compute::NormaliseGradient() to handle optimise* parameters

onurulgen committed:
Move platform and measure initialisation into reg_base::SetPlatformType()

onurulgen committed:
Refactor NR_CUDA_SAFE_CALL() and NR_CUDA_CHECK_KERNEL()

onurulgen committed:
Refactorisations

onurulgen committed:
Modernise CUDA conjugate gradient functions

- Ditch old texture objects and use up-to-date ones
- Make texture objects managed
- Ditch CUDA symbols and pass them as kernel function parameters
- Extend reg_updateControlPointPosition_gpu() to handle optimise* parameters

onurulgen committed:
Remove unnecessary testing functions

onurulgen committed:
Add tests for conjugate gradient

onurulgen committed:
Refactorisations

 - Rename CudaContextSingleton as CudaContext, and move it into NiftyReg namespace
 - Rename NiftyReg_CudaBlock100 as BlockSize, and move it into NiftyReg namespace
 - Move BlockSize implementation into the header
 - Change the type of BlockSize members as unsigned
 - Move BlockSize instance into CudaContext
 - Use unsigned instead of size_t in CUDA kernels
 - Initialise the CUDA or OpenCL device in Platform's constructor
 - Rename `unsigned int`s as `unsigned`

onurulgen committed:
Generate test data only once for NormaliseGradientTest

onurulgen committed:
Refactorisations

onurulgen committed:
Add *Content::UpdateWarped() to transfer warped image to the CUDA device

onurulgen committed:
Use CPU for GetWarpedImage()

onurulgen committed:
Use CPU for reg_f3d2::GetWarpedImage()

onurulgen committed:
Handle NiftiImage copy options with an enum

onurulgen committed:
Fix GPU version of ResampleImage() to make on a par with CPU version

onurulgen committed:
Fix GPU version of GetImageGradient() to make on a par with CPU version

onurulgen committed:
Refactor cudaCommon_createTextureObject()

onurulgen committed:
Move EPS value into reg_test_common.h

onurulgen committed:
Refactor reg_localTransformation_kernels.cu

onurulgen committed:
Refactor reg_localTransformation_gpu.cu

onurulgen committed:
Fix GPU version of GetDeformationField() to make on a par with CPU version

onurulgen committed:
Add tests for *Compute::GetDeformationField()

onurulgen committed:
Refactorisations

onurulgen committed:
Add tests for *Compute::VoxelCentricToNodeCentric()

onurulgen committed:
Fix GPU version of VoxelCentricToNodeCentric() to make on a par with CPU version

onurulgen committed:
Fix macOS compilation errors

onurulgen committed:
Fix the precision bug of reg_lncc

onurulgen committed:
Optimise reg_voxelCentric2NodeCentric_kernel()

mmodat committed:
Issue #92: Added LNCC unit tests

mmodat committed:
Merge branch 'rearchitect-and-reinstate-gpu' of https://github.com/KCL-BMEIS/niftyreg into rearchitect-and-reinstate-gpu

onurulgen committed:
Refactorisations

onurulgen committed:
Fix a bug causing wrong calculation of the affine transformation matrix

mmodat committed:
Issue #92: fix affine initialisation in f3d. matrix went out of scope during refactoring.

onurulgen committed:
Copy affine transformation into reg_base instead of linking

onurulgen committed:
Make affine transformation memory managed in reg_aladin*

onurulgen committed:
Fix CUDA 12 incompatibilities

onurulgen committed:
Fix a memory leak

mmodat committed:
#92: block match & LTS CPU/CUDA regression tests

onurulgen committed:
Refactor the LTS regression test

onurulgen committed:
Refactor the BlockMatching regression test

onurulgen committed:
Refactor the LNCC test

onurulgen committed:
Rename OptimiseKernel to LtsKernel

mmodat committed:
Issue#92: Added BM unit test and fix mask handling

onurulgen committed:
Refactorisations

mmodat committed:
#92 fixed shared mem race condition in block match

onurulgen committed:
Add NiftiImage::getDataType() returning the current data type

onurulgen committed:
Refactor _reg_resampling

onurulgen committed:
Fix memory leaks in Content

onurulgen committed:
Refactor reg_nmi

onurulgen committed:
Add symmetric scheme support for reg_nmi_gpu #92

onurulgen committed:
Add NMI regression tests #92

mmodat committed:
#92: Set the test seed to constant for reproducibility

mmodat committed:
Issue #92: added a unit test for nmi

mmodat committed:
Issue #92: activate CUDA for NMI test

mmodat committed:
#72 #92 Added an error message when ln is set to 0

onurulgen committed:
Add symmetric scheme support for reg_optimiser_gpu and reg_conjugateGradient_gpu #92

onurulgen committed:
Add symmetric scheme support for reg_test_conjugateGradient #92

mmodat committed:
#92: added bending energy tests

mmodat committed:
Issue #68: checks for exising folder before saving

onurulgen committed:
Refactorisations

onurulgen committed:
Rearchitect reg_measure to handle forward and backward similarity measure values #92

onurulgen committed:
Disable OpenMP for coverage

onurulgen committed:
Rearchitect reg_measure to handle forward and backward voxel-based similarity measure gradient computation #92

- Add symmetric scheme support for reg_ssd_gpu

onurulgen committed:
Remove the symmetric scheme warning from reg_ssd_gpu

onurulgen committed:
Rearchitect Content classes #92

onurulgen committed:
Add Debug header to handle errors and logging #92

 - Handle errors and throw exception instead of exiting the app
 - Print errors, warnings and infos with NR_ERROR, NR_WARN and NR_INFO macros respectively
 - Print verbose messages with NR_VERBOSE
 - Print function names with NR_FUNC_CALLED() macro
 - Convert all *printf to NR_COUT or NR_CERR
 - Get rid of NDEBUGs and use NR_DEBUG instead
 - Replace C-style string manipulation with C++-style
 - Ditch reg_stringFormat()

onurulgen committed:
Refactor _reg_common_cuda #92

onurulgen committed:
Convert NMI regression test to a multi-measure regression test #92

onurulgen committed:
Refactor Cuda::TransferNiftiToDevice and remove Cuda::TransferNiftiToDeviceSimple

onurulgen committed:
Make SSD GPU on a par with the CPU version #92

mmodat committed:
#92 Added nmi gradient test

mmodat committed:
#92 Added nmi gradient test + pull

mmodat committed:
#92 removed readwritebinary as not needed

mmodat committed:
#92 changed the def field test to be a unit test against known output

mmodat committed:
#92 Added spine composition to the unit test

onurulgen committed:
Add NiftiImage::setPixDim()

onurulgen committed:
Fix test errors

onurulgen committed:
Turn math macros into inline functions #92

mmodat committed:
#92 addded test for composition of DF. CPU only for now

onurulgen committed:
Fix a bug causing shrunk output images

onurulgen committed:
Refactor Compute::GetApproximatedGradient()

onurulgen committed:
Refactorisations

onurulgen committed:
Add NiftiImage::setIntentName()

onurulgen committed:
Add more float* operations

onurulgen committed:
Add CUDA image operations

onurulgen committed:
Add reg_getMinValue_gpu() and reg_getMaxValue_gpu()

onurulgen committed:
Implement CudaCompute::GetDefFieldFromVelocityGrid() #92

mmodat committed:
#92: fix issue in reg_resample - fix failing test - fix output display

onurulgen committed:
Merge branch 'rearchitect-and-reinstate-gpu' of https://github.com/KCL-BMEIS/niftyreg into rearchitect-and-reinstate-gpu

onurulgen committed:
Fix a bug causing early freeing of the image data

Passing a nifti_image pointer to a NiftiImage causes unwanted freeing of the nifti_image pointer

onurulgen committed:
Refactorisations

onurulgen committed:
Fix precision of GetBasisSplineValue() and GetBasisSplineDerivativeValue()

onurulgen committed:
Fix a bug causing inconsistent results in successive runs of 2D f3d registration #92

onurulgen committed:
Refactorisations

onurulgen committed:
Fix wrong calculation of CUDA ApproxLinearEnergyGradient(), GetLandmarkDistance(), and LandmarkDistanceGradient() #92

onurulgen committed:
Refactorisations

onurulgen committed:
Implement CudaCompute::ApproxLinearEnergyGradient() #92

onurulgen committed:
Fixes for CUDA 12

onurulgen committed:
Add approximate linear energy gradient regression test #92

onurulgen committed:
Implement CudaCompute::ApproxLinearEnergy() #92

Also refactor CudaCompute::ApproxLinearEnergyGradient()

onurulgen committed:
Add approximate linear energy regression test #92

onurulgen committed:
Fix a bug in reg_spline_approxLinearEnergyGradient_gpu()

onurulgen committed:
Refactorisations

onurulgen committed:
Implement the kernel convolution for CUDA #92

onurulgen committed:
Add regression tests for kernel convolution #92

onurulgen committed:
Use ConvKernelType instead of NREG_CONV_KERNEL_TYPE

onurulgen committed:
Fix a bug in the kernel convolution regression test

onurulgen committed:
Optimise Cuda::KernelConvolution()

onurulgen committed:
Optimise reg_getMinMaxValue_gpu()

onurulgen committed:
Update compose deformation field test to include CUDA

onurulgen committed:
Add composition support for CudaCompute::GetDeformationField() #92

onurulgen committed:
Refactorisations

onurulgen committed:
Enable CUDA for NMI gradient unit test

onurulgen committed:
Add content creator for f3d2 #92

This fixes incorrect NMI calculation for CUDA by sharing reference and floating CUDA arrays between contents

onurulgen committed:
Refactorisations

onurulgen committed:
Add regression tests for Compute::GetDeformationField() #92

onurulgen committed:
Remove identifiers starting with a single underscore

Identifiers in global scope starting with a single underscore are reserved and isn't allowed to use

onurulgen committed:
Refactorisations

onurulgen committed:
Fix a bug in CudaCompute::VoxelCentricToNodeCentric() #92

onurulgen committed:
Make CudaCompute::ResampleImage() on a par with CPU #92

onurulgen committed:
Make CudaCompute::GetDeformationField() on a par with CPU #92

onurulgen committed:
Temporarily disable GetDeformationFieldTest

onurulgen committed:
Make CudaCompute::NormaliseGradient() on a par with CPU #92

onurulgen committed:
Optimise CudaCompute::GetMaximalLength() #92

onurulgen committed:
Update tests

onurulgen committed:
Make CudaCompute::ApproxBendingEnergyGradient() on a par with CPU #92

onurulgen committed:
Add regression tests for approximate bending energy and approximate bending energy gradient #92

onurulgen committed:
Refactorisations

onurulgen committed:
Implement reg_getNmiValue for CUDA #92

onurulgen committed:
Make reg_getVoxelBasedNmiGradient_gpu() on a par with CPU #92

- Optimise reg_getVoxelBasedNmiGradient_gpu()
- Get the function ready for multi-timepoint support

onurulgen committed:
Add multi-timepoint support for MeasureTest #92

onurulgen committed:
Refactorisations

onurulgen committed:
Convert reference and floating images to float arrays from cudaArrays #92

 - Eliminate unnecessary Cuda::* functions
 - Refactor Cuda::CreateTextureObject()

onurulgen committed:
Refactorisations

onurulgen committed:
Implement multi-timepoint support for CUDA #92

onurulgen committed:
Implement reg_optimiser_gpu::Perturbation() #92

onurulgen committed:
Refactor Optimiser #92

onurulgen committed:
Optimise Optimiser #92

onurulgen committed:
Refactorisations

onurulgen committed:
Use float gam instead of double in CudaOptimiser

Even though tests show otherwise, using float gets better results in real world scenarios.

onurulgen committed:
Optimise CudaResampling #92

onurulgen committed:
Combine GitHub Actions workflows for testing into a single file

onurulgen committed:
Update GitHub Actions workflow for coverage

onurulgen committed:
Refactor CudaLocalTransformation

onurulgen committed:
Optimise Cuda::ApproxLinearEnergyGradient() #92

onurulgen committed:
Refactorisations

onurulgen committed:
Optimise Cuda::GetDeformationField() #92

onurulgen committed:
Refactor CudaTools #92

onurulgen committed:
Optimise Cuda::VoxelCentricToNodeCentric() #92

onurulgen committed:
Optimise Cuda::DefFieldCompose() #92

onurulgen committed:
Implement CudaCompute::SymmetriseVelocityFields() #92

onurulgen committed:
Add symmetrise velocity fields regression test #92

onurulgen committed:
Implement CudaCompute::UpdateVelocityField() #92

onurulgen committed:
Add update velocity field regression test #92

onurulgen committed:
Print the version info

onurulgen committed:
Implement Cuda::ResampleGradient() #92

onurulgen committed:
Add resample gradient regression test #92

onurulgen committed:
Refactorisations

onurulgen committed:
Add Cuda::GetIntermediateDefFieldFromVelGrid() #92

onurulgen committed:
Implement CudaCompute::ExponentiateGradient() #92

onurulgen committed:
Add exponentiate gradient regression test #92

onurulgen committed:
Refactor Cuda::GetAffineDeformationField() #92

onurulgen committed:
Refactor affine deformation field unit test #92

onurulgen committed:
Add composition to affine deformation field unit test #92

onurulgen committed:
Add testing for compute to affine deformation field unit test #92

onurulgen committed:
Use real index numbers returned from maskCuda in deformationFieldCuda and warpedGradientCuda #92

onurulgen committed:
Use Codecov for coverage

onurulgen committed:
Update minimum CMake version and use the new CUDA method

onurulgen committed:
Add a GitHub Action to build executables on a release #92

onurulgen committed:
Use self-hosted runners to enable CUDA for coverage

onurulgen committed:
Add a GitHub Action for static code analysis

onurulgen committed:
Fix GPU ID parsing error in reg_f3d

onurulgen committed:
Upgrade libpng to 1.6.42

onurulgen committed:
Upgrade zlib to 1.3.1

onurulgen committed:
Upgrade RNifti and dependencies

onurulgen committed:
Fix zlib compilation error

onurulgen committed:
Fix linting issues of reg_png

onurulgen committed:
Fix linting issues of nifti1_io

onurulgen committed:
Enable inline suppressions for static code analysis

onurulgen committed:
Exclude Eigen library from static code analysis

onurulgen committed:
Enable CRT secure warnings

onurulgen committed:
Fix linting issues

onurulgen committed:
Eliminate dead code

onurulgen committed:
Use git to download Eigen library

Also, upgrade Eigen library to 3.3.*

onurulgen committed:
Refactorisations

onurulgen committed:
Rename Measure as MeasureCreator

onurulgen committed:
Refactor _reg_maths* and _reg_common_cuda_kernels

onurulgen committed:
Embed Eigen library

onurulgen committed:
Replace Divide() with IndexToDims()

onurulgen committed:
Add include directories for OpenMP

onurulgen committed:
Refactorisations

onurulgen committed:
Update macOS images in GitHub Actions

onurulgen committed:
Update the OS image for Coverage

onurulgen committed:
Fix linting issues

onurulgen committed:
Use NiftiImage instead of nifti_image in _reg_ReadWriteImage

Delete duplicate reg_io_ReadImageHeader() and add this ability to reg_io_ReadImageFile()

onurulgen committed:
Change the image acquisition behaviour of NiftiImage

Don't own the image pointer if it's constructed by using a nifti_image pointer

onurulgen committed:
Revert "Enable CRT secure warnings"

This reverts commit 6cbbccd4d1452ad7870126a3f455164888dcd703.

onurulgen committed:
Fix *Content::CastImageData() and eliminate its duplicates

onurulgen committed:
Use NiftiImage in Content classes

onurulgen committed:
Revert "Use float gam instead of double in CudaOptimiser"

This reverts commit b9c9beca65c9c7f6862e1c1ca50c70eaafd1fbfa.

onurulgen committed:
Fix Floor(), Ceil(), and Round() functions

onurulgen committed:
Merge branch 'master' into rearchitect-and-reinstate-gpu

msseibel commented:
I had to add `set(CMAKE_CUDA_ARCHITECTURES "native")` to the CMakeLists.txt file. Otherwise, I got the following error, "CMAKE_CUDA_ARCHITECTURES must be non-empty if set". 

Besides that, the build worked, and I was able to run the program.

onurulgen commented:
> I had to add `set(CMAKE_CUDA_ARCHITECTURES "native")` to the CMakeLists.txt file. Otherwise, I got the following error, "CMAKE_CUDA_ARCHITECTURES must be non-empty if set".
> 
> Besides that, the build worked, and I was able to run the program.

CMAKE_CUDA_ARCHITECTURES are set in [here](https://github.com/KCL-BMEIS/niftyreg/blob/da1da73935f5b4871aa6121955d1040ce5de5b31/reg-lib/cuda/CMakeLists.txt). If you need to set that, your _cmake configure_ command must've failed while detecting CUDA, right?

msseibel commented:
Yes, that is correct. It failed as soon as it reached line 156 `enable_language(CUDA)`.

onurulgen committed:
Fix typos

onurulgen committed:
Fix performance bugs

onurulgen committed:
Don't set build type for coverage but warn

onurulgen committed:
Use Git commit hash as the build number

onurulgen committed:
Remove build type checks for non-MSVC environments

onurulgen committed:
Update macOS version in CI/CD workflows to macOS 13

onurulgen committed:
Skip calculation of cal_min and cal_max in NiftiImage::replaceData()

onurulgen committed:
Fix min calculation in NiftiImageData::ConcreteTypeHandler to use lowest() for numeric limits

onurulgen committed:
Fix min/max calculation in NiftiImageData::ConcreteTypeHandler to use magnitudes for complex numbers

onurulgen committed:
Refactor minmax methods in NiftiImageData to return std::pair<double, double> instead of using output parameters

onurulgen committed:
Remove internal lessThan function and simplify min/max comparisons in NiftiImageData::ConcreteTypeHandler

onurulgen committed:
Remove unused string variable in main function of reg_f3d

onurulgen committed:
Handle scaling in NiftiImageData::minmax()

onurulgen committed:
Simplify minmax calculations in NiftiImageData::ConcreteTypeHandler by using std::minmax_element

onurulgen committed:
Refactor squaring number calculations in reg_defField_getDeformationFieldFromFlowField

onurulgen committed:
Refactor min/max value retrieval to use NiftiImageData.minmax() and remove deprecated reg_tools_getMinValue/MaxValue functions

onurulgen committed:
Add .cache to .gitignore

onurulgen committed:
Refactor squaring number calculations in Cuda::GetDeformationFieldFromFlowField()

onurulgen committed:
Refactor min/max value retrieval to use a unified GetMinMaxValue function

onurulgen committed:
Add VSCode Dev Containers to ease development

onurulgen committed:
Add .clang-format configuration for consistent code styling

onurulgen committed:
Update versioning to use Git tags

onurulgen committed:
Update GitHub Actions workflows

onurulgen committed:
Fix Cppcheck installation in Code Analysis workflow

onurulgen committed:
Fix runner label in Coverage workflow

kaczmarj commented:
hi @onurulgen - thank you for your tremendous effort in getting CUDA support back into niftyreg for a variety of realignment methods.

is this branch OK to try out? i am currently using version 1.3.9 from 2012 for `reg_f3d_gpu`.

onurulgen commented:
> hi @onurulgen - thank you for your tremendous effort in getting CUDA support back into niftyreg for a variety of realignment methods.
> 
> is this branch OK to try out? i am currently using version 1.3.9 from 2012 for CUDA-enabled `reg_f3d`.

Yes, it's stable, you may go ahead and test it. I'd like to hear how it goes as well. Could you please give some feedback if you can?

onurulgen committed:
Fix compilation errors

onurulgen committed:
Update release workflow for Windows and CMake flags

Switches deprecated Windows runner from 2019 to 2022. Updates CMake flags for Windows and macOS builds to fix compilation errors.

kaczmarj commented:
@onurulgen - niftyreg v2.0.0 is working great so far. thanks again for all of your work.

i compiled it in a docker container. i copied the dockerfile below if anyone is interested. i got errors with CUDA 12.x so i stuck with 11.8

```dockerfile
FROM nvcr.io/nvidia/cuda:11.8.0-devel-ubuntu22.04 AS builder

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y \
        build-essential \
        cmake \
        git \
        ninja-build \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*

WORKDIR /work
RUN git clone https://github.com/KCL-BMEIS/niftyreg . \
    && git checkout v2.0.0

RUN cmake -S . -B build -G "Ninja" \
        -DCMAKE_BUILD_TYPE=Release \
        -DBUILD_ALL_DEP=ON \
        -DUSE_CUDA=ON \
        -DCHECK_GPU=OFF \
        -DUSE_OPENMP=ON \
        -DUSE_SSE=ON \
        -DCMAKE_INSTALL_PREFIX=/opt/niftyreg \
        -DCMAKE_INSTALL_RPATH="\$ORIGIN/../lib" \
        # Support many different NVIDIA GPUs.
        -DCMAKE_CUDA_ARCHITECTURES="75-real;80-real;86-real;89-real;90" \
    && cmake --build build --parallel \
    && cmake --install build

FROM nvcr.io/nvidia/cuda:11.8.0-runtime-ubuntu22.04
COPY --from=builder /opt/niftyreg /opt/niftyreg
ENV PATH="/opt/niftyreg/bin:${PATH}"
WORKDIR /work
```

i built with this (i'm on an arm-based macbook)

```bash
docker buildx build --platform=linux/amd64 --load -t kaczmarj/niftyreg .
```

then I got the binaries out with this:

```bash
docker run --rm \
  -v "$(pwd)":/out \
  kaczmarj/niftyreg \
  bash -c "cd /opt && tar -czf /out/niftyreg.tar.gz niftyreg"
```

onurulgen commented:
Thank you! You may use [precompiled binaries](https://github.com/KCL-BMEIS/niftyreg/releases); and if you'd like to compile yourself, you may follow [release.yml](https://github.com/KCL-BMEIS/niftyreg/blob/master/.github/workflows/release.yml) file. The code is compatible up to CUDA 12.4.x. It might've failed if you tried later versions.

kaczmarj commented:
fantastic, i will use the precompiled binaries in that case. thanks kindly
